plugins/dfu/dfu-target-stm.c

```diff
@@ -0,0 +1,496 @@
+		sector = dfu_target_get_sector_for_addr (target, offset_dev);
+		g_assert (sector != NULL);
+
+		/* manually set the sector address */
+		if (dfu_sector_get_zone (sector) != zone_last) {
+			g_debug ("setting address to 0x%04x",
+				 (guint) offset_dev);
+			if (!dfu_target_stm_set_address (target,
+							   (guint32) offset_dev,
+							   cancellable,
+							   error))
+				return FALSE;
+			zone_last = dfu_sector_get_zone (sector);
+		}
+
+		/* we have to write one final zero-sized chunk for EOF */
+		length = g_bytes_get_size (bytes) - offset;
+		if (length > transfer_size)
+			length = transfer_size;
+		bytes_tmp = g_bytes_new_from_bytes (bytes, offset, length);
+		g_debug ("writing sector at 0x%04x (0x%" G_GSIZE_FORMAT ")",
+			 offset_dev,
+			 g_bytes_get_size (bytes_tmp));
+		/* ST uses wBlockNum=0 for DfuSe commands and wBlockNum=1 is reserved */
+		if (!dfu_target_download_chunk (target,
+						(guint8) (i + 2),
+						bytes_tmp,
+						cancellable,
+						error))
+			return FALSE;
+
+		/* getting the status moves the state machine to DNLOAD-IDLE */
+		if (!dfu_target_check_status (target, cancellable, error))
+			return FALSE;
+
+		/* update UI */
+		dfu_target_set_percentage (target, offset, g_bytes_get_size (bytes));
+	}
+
+	/* done */
+	dfu_target_set_percentage_raw (target, 100);
+	dfu_target_set_action (target, FWUPD_STATUS_IDLE);
+
+	/* success */
+	return TRUE;
+}
+
+#if 0
+static gboolean
+dfu_target_stm_get_commands (DfuTarget *target,
+			     GCancellable *cancellable,
+			     GError **error)
+{
+	GBytes *data_in;
+	GBytes *data_out;
+	guint8 buf[1];
+
+	/* format buffer */
+	buf[0] = DFU_STM_CMD_GET_COMMAND;
+	data_in = g_bytes_new_static (buf, sizeof(buf));
+	if (!dfu_target_download_chunk (target, 0, data_in, cancellable, error)) {
+		g_prefix_error (error, "cannot get DfuSe commands: ");
+		return FALSE;
+	}
+
+	/* return results */
+	data_out = dfu_target_upload_chunk (target, 0, cancellable, error);
+	if (data_out == NULL)
+		return FALSE;
+
+	// N bytes,
+	// each byte is the command code
+
+	// FIXME: parse?
+	return TRUE;
+}
+
+/**
+ * dfu_target_stm_read_unprotect:
+ * @target: a #DfuTarget
+ * @cancellable: a #GCancellable, or %NULL
+ * @error: a #GError, or %NULL
+ *
+ * Turns of read protection on the device, clearing all SRAM and EEPROM memory.
+ *
+ * Return value: %TRUE for success
+ **/
+static gboolean
+dfu_target_stm_read_unprotect (DfuTarget *target,
+			       GCancellable *cancellable,
+			       GError **error)
+{
+	GBytes *data_in;
+	guint8 buf[5];
+
+	/* format buffer */
+	buf[0] = DFU_STM_CMD_READ_UNPROTECT;
+	memcpy (buf + 1, &address, 4);
+	data_in = g_bytes_new_static (buf, sizeof(buf));
+	if (!dfu_target_download_chunk (target, 0, data_in, cancellable, error)) {
+		g_prefix_error (error, "cannot read-unprotect: ");
+		return FALSE;
+	}
+
+	/* for ST devices, the action only occurs when we do GetStatus */
+	return dfu_target_check_status (target, cancellable, error);
+}
+#endif
```

Comment:
Why is this stuff commented out?  If it's going to stick around, at least indicate what the TODO is to restore it.

Reply:
I guess it can go completely, we have git history in case it's ever required. Will Fix.